client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java

```diff
@@ -19,35 +19,158 @@
+            throw e;
         }
+        return responseConverter.apply(response);
     }
 
+    private <Req extends ActionRequest, Resp> void performRequestAsyncAndParseEntity(Req request, Function<Req, Request> requestConverter,
+            CheckedFunction<XContentParser, Resp, IOException> entityParser, ActionListener<Resp> listener, Header... headers) {
+        performRequestAsync(request, requestConverter, (response) -> parseEntity(response.getEntity(), entityParser), listener, headers);
+    }
+
+    private <Req extends ActionRequest, Resp> void performRequestAsync(Req request, Function<Req, Request> requestConverter,
+            CheckedFunction<Response, Resp, IOException> responseConverter, ActionListener<Resp> listener, Header... headers) {
+        ActionRequestValidationException validationException = request.validate();
+        if (validationException != null) {
+            listener.onFailure(validationException);
+        }
+        Request req = requestConverter.apply(request);
+        ResponseListener responseListener = wrapResponseListener(responseConverter, listener);
+        client.performRequestAsync(req.method, req.endpoint, req.params, req.entity, responseListener, headers);
+    }
+
+    static <Resp> ResponseListener wrapResponseListener(CheckedFunction<Response, Resp, IOException> responseConverter,
+                                                        ActionListener<Resp> actionListener) {
+        return new ResponseListener() {
+            @Override
+            public void onSuccess(Response response) {
+                try {
+                    actionListener.onResponse(responseConverter.apply(response));
+                } catch(Exception e) {
+                    onFailure(e);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception exception) {
+                /* TODO add parsing of root exception, printed out via BytesRestResponse#convert
+                if (exception instanceof ResponseException) {
+                    HttpEntity entity = ((ResponseException) exception).getResponse().getEntity();
+                    try {
+                        ElasticsearchException elasticsearchException = parseEntity(request, entity, here goes the parsing method ref);
+                        actionListener.onFailure(elasticsearchException);
+                        return;
+                    } catch (IOException e) {
+                        exception.addSuppressed(e);
+                    }
+                }*/
+                actionListener.onFailure(exception);
+            }
+        };
+    }
 
+    static <Resp> Resp parseEntity(
+            HttpEntity entity, CheckedFunction<XContentParser, Resp, IOException> entityParser) throws IOException {
+        if (entity == null) {
+            throw new IllegalStateException("Response body expected but not returned");
```

Comment:
I wanted to highlight that it may be fine to not get a response body in some cases (not all api return one), but in this case it was expected hence we are in trouble.